runtime/examples/Rust/tapasco-debug/src/app/mod.rs

```diff
@@ -0,0 +1,336 @@
+use std::{
+    collections::HashMap,
+};
+
+use tapasco::{
+    tlkm::*,
+    device::status::Interrupt,
+    pe::PE, 
+    device::PEParameter,
+};
+
+use chrono::{TimeZone, Utc};
+
+use snafu::{ResultExt, Snafu};
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Failed to initialize TLKM object: {}", source))]
+    TLKMInit { source: tapasco::tlkm::Error },
+
+    #[snafu(display("Failed to decode/acquire TLKM device or one of its PEs: {}", source))]
+    DeviceInit { source: tapasco::device::Error },
+}
+
+type Result<T, E = Error> = std::result::Result<T, E>;
+
+
+// TODO: Right now I'm just handling the first FPGA device because I assume most will have
+// only one FPGA but switching devices should be possible e.g. by passing the device as cli
+// argument or with an extra key binding?
+// TODO: It would be nice to see when/how many interrupts were triggered.
+
+
+pub struct App<'a> {
+    bitstream_info: String,
+    platform_info: String,
+    pub title: &'a str,
+    pub tabs: TabsState<'a>,
+    pub pe_infos: StatefulList<String>,
+    pub pes: Vec<(usize, PE)>, // Plural of Processing Element (PEs)
+}
+
+impl<'a> App<'a> {
+    pub fn new() -> Result<App<'a>> {
+        // Get Tapasco Loadable Linux Kernel Module
+        let tlkm = TLKM::new().context(TLKMInit {})?;
+        //let tlkm_device = tlkm.device_enum(&HashMap::new()).context(TLKMInit {})?;
+        // Allocate the first device with ID 0 (because most devices will only have one FPGA)
+        let mut tlkm_device = tlkm.device_alloc(0, &HashMap::new()).context(TLKMInit {})?;
+        // Change device access to excluse to be able to acquire PEs
+        tlkm_device.change_access(tapasco::tlkm::tlkm_access::TlkmAccessExclusive).context(DeviceInit {})?;
```

Comment:
This will be an issue as you cannot run any other tapasco software requiring exclusive access while running tapasco-debug this way.

You might want to change your "acquire_pe_without_job" to return a read-only version of the PE which does not require exclusive access.

Reply:
Yes, I wanted to discuss this anyway, so I've split it into a separate commit. Thank you for bringing it up directly.

I'm not sure how to resolve this in the end but I think in order to set register values (which is not implemented at the moment) there should be exclusive access. If someone wants to observe another Tapasco application there could be a read-only mode of `tapasco-debug` where you cannot set the values but run with other software side-by-side.

What do you think about this?

Reply:
We might want to have ´tapasco-debug´ ignore all of the scheduler etc and go entirely with debug functionality. This could look like the functions I've prepared for this purpose for the platform part at https://github.com/esa-tu-darmstadt/tapasco/blob/50ed109d1fb2b6354fbe3eede8804b5066365a94/runtime/libtapasco/src/device.rs#L618

For the Arch part we'd only need two functions, e.g.:
```Rust
pub fn get_available_pes(&self) -> Vec<(Id, PEId)> {}

pub fn get_pes(&self, read_only: bool) -> PE {}
```

For those we can have a read-only version which does not require exclusive access, and a full access version. Tapasco-debug can then have e.g. three modes
 - Default: Monitor mode -> read only
 - Full Access -> requires exclusive
 - Forced Full Access -> Works with monitor mode but provides full access, all safety is off

Reply:
Sounds good to me. I'm going to talk about it with @cahz later and see what his opinion is on this.

Reply:
There are now three modes: Monitor, Debug and Unsafe. But they don't work correctly yet. See:
https://github.com/esa-tu-darmstadt/tapasco/issues/296

Reply:
I have now disabled the debug mode again and added a comment for #296.